### ISSUE TITLE
INNER JOIN becomes LEFT OUTER JOIN when merging relations

### ISSUE THREAD

**Issue by joanniclaborde** [rails/rails#16140]
Rails: 4.0.6
Database: MySQL

Given the following tables and models: https://gist.github.com/joanniclaborde/26702fc1ee80c74d315e

Getting T2s with T3s generates an `INNER JOIN` query as expected:

```
> T2.with_t3.to_sql
 => "SELECT `t2s`.* FROM `t2s` INNER JOIN `t3s` ON `t3s`.`id` = `t2s`.`t3_id`"
```

But merging that relation with another one turns the join into a `LEFT OUTER JOIN`:

```
> T1.joins(:t2).merge(T2.with_t3).to_sql
 => "SELECT `t1s`.* FROM `t1s` INNER JOIN `t2s` ON `t2s`.`id` = `t1s`.`t2_id` LEFT OUTER JOIN `t3s` ON `t3s`.`id` = `t2s`.`t3_id`" 
```

I'm not sure if it is the expected behaviour or not? I fixed it in my T2 model with a custom join:

```
class T2 < ActiveRecord::Base
  scope :with_t3, -> { joins('INNER JOIN t3s ON t3s.id = t2s.t3_id') }
end
```

**Comment by gamov** [comment]
I've noticed this happening with my merges since Rails 4.0.
I don't understand @thedarkone idea to force an inner join. Do you mean:
```ruby
T1.joins(:t2 => :t3).merge(T2.with_t3)
```
In my case, this doesn't work, it **adds** an inner join but the merge outer join is still there, at the end.
Like @joanniclaborde, I also love to think of scopes as beautifully encapsulated black box so adding a join here break the idea and is an implementation detail of the scope.

I've also seen some cases of wrong ordering of joins since Rails 4.0, e.g.:
```sql
SELECT 
    `container_item_allocations`.*
FROM
    `container_item_allocations`
        INNER JOIN
    `shipping_items` ON `shipping_items`.`id` = `container_item_allocations`.`shipping_item_id`
    -- Notice how the following two joins are reversed this the query is invalid
        LEFT OUTER JOIN
    `item_families` ON `item_families`.`id` = `item_variants`.`item_family_id`
        LEFT OUTER JOIN
    `item_variants` ON `item_variants`.`id` = `shipping_items`.`item_variant_id`
```

**Comment by thedarkone** [comment]
@gamov as far as I understood, @joanniclaborde wanted to `SELECT t1 INNER JOIN t2 INNER JOIN t3`, so I suggested to use `T1.joins(:t2 => :t3)` as it should generate a query exactly as requested. Your example with tacked on `.merge(T2.with_t3)` doesn't work, because it tacks on `.merge(T2.with_t3)` at the end.

I'm not defending Rails' behavior, I'm trying to suggest a workable solution.

re your example with swapped joins, I have #15461 that attempts to fix this. The PR is a bit stale now, but I'll try to look into it really soon.

**Comment by MaxLap** [comment]
I searched a bit and never found a rationale for this behaviour. 

So I went git-hunting for when the behavior appeared, to maybe find the rationale there, or find person that could shine light on this question. 

TLDR: I don't think there is any rationale for this. It's just an oversight that was made when a feature was improved. (Back in rails 3.2, you couldn't do merges with joins like that)

This is the pull request that allowed doing the merge of joins in a coherent way https://github.com/rails/rails/pull/10164. Before that, doing a merge with a joins that is not on the initial model failed (ex: `Post.joins(:comments).merge(Comment.joins(:ratings))`). 

Looking at the code, I don't think there is any rationale. It's simply an oversight. The code used in that initial commit is the exact same one used for merging joins and includes.
https://github.com/rails/rails/blob/dc764fcc348d562376add26ff8ef5173946b575b/activerecord/lib/active_record/relation/merger.rb#L97-L99
https://github.com/rails/rails/blob/dc764fcc348d562376add26ff8ef5173946b575b/activerecord/lib/active_record/relation/finder_methods.rb#L230-L232

Those 2 identical locations where later merged into one method in https://github.com/rails/rails/commit/fba18f19948f084023fd8744025f57da00163265

Then, tenderlove did a pretty massive refactor of this zone in october 2013, keeping the behavior as it was. 

I will probably try to do a PR to solve this issue when I get the time.
